### PR TITLE
[RISCV] Enable the TypePromotion pass from AArch64/ARM.

### PR DESCRIPTION
**Pull request by topperc**
This pass looks for unsigned icmps that have illegal types and tries
to widen the use/def graph to improve the placement of the zero
extends that type legalization would need to insert.
    
I've explicitly disabled it for i32 by adding a check for
isSExtCheaperThanZExt to the pass.
    
The generated code isn't perfect on the lit test, but my data shows a net
dynamic instruction count improvement on spec2017 for both base and
Zba+Zbb+Zbs.

Spec2017 dynamic count for rv64gc_zba_zbb_zbs

| Metric |Before | After | Before/After |
|-------|------|-----|------|
| 500.perlbench_r | 3032622956407 | 3032363566062 | 99.991% |
| 502.gcc_r | 1231223971793 | 1226644351851 | 99.628% |
| 505.mcf_r | 696461864012 | 696461368989 | 100.000% |
| 520.omnetpp_r | 1142136357540 | 1142234473240 | 100.009% |
| 523.xalancbmk_r | 1157112062701 | 1148472443956 | 99.253% |
| 525.x264_r | 3970723342408 | 3970671717801 | 99.999% |
| 531.deepsjeng_r | 1833000391081 | 1833000391091 | 100.000% |
| 541.leela_r | 2059859108461 | 2059858783398 | 100.000% |
| 557.xz_r | 1806198505643 | 1806198505653 | 100.000% |

### CODE REVIEW

**Comment by dtcxzyw** [comment]
> The generated code isn't perfect on the lit test, but my data shows a net
dynamic instruction count improvement on spec2017 for both base and
Zba+Zbb+Zbs.

I agree.

File: llvm/test/CodeGen/RISCV/signbit-test.ll [hunk 1]
```diff
@@ -312,7 +315,10 @@ define i16 @test_clear_mask_i16_i8(i16 %x) nounwind {
 ; RV64-NEXT:    bnez a1, .LBB10_2
 ; RV64-NEXT:  # %bb.1: # %t
 ; RV64-NEXT:    li a0, 42
-; RV64-NEXT:  .LBB10_2: # %f
+; RV64-NEXT:    ret
+; RV64-NEXT:  .LBB10_2:
+; RV64-NEXT:    slli a0, a0, 48
```
Comment (preames) [inline llvm/test/CodeGen/RISCV/signbit-test.ll:320]: This delta is interesting and looks like a missed optimization.  After moving the zext, we appear to loose track of the fact that x was already used in the ABI, and thus should have the required properties for the return.

**Review by preames — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/RISCV/signbit-test.ll (hunk 1 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/signbit-test.ll:320]: We moved the zext to the argument and updated both uses. Then in SelectionDAG we type legalize that to an AND. There's already an AND with 128 after it going to the compare so we decide we don't need the other AND on that path. That leaves the AND only serving the return which doesn't care about the upper bits.

Seems like there was no reason to move the zext for this case.

**Comment by topperc** [comment]
Committed as 9838c8512bc29e3a1b8edeb0eb2541160e4c727f and 7d40ea85d5ea5cc837536f61e3b4f80ea69f14d0

**Pull request closed without merging**